### PR TITLE
Reveal ballots and count votes

### PR DESCRIPTION
**Pull request by akuanti**
This PR implements ballot revealing and vote counting, including runoff vote counting.

If after the first vote count, one slate has at least 50%, the contest's status is set to `VoteFinalized`. Otherwise, the contest's status is set to `RunoffPending`, and a runoff can be triggered. In the runoff, all slates other than the top two from the first count are eliminated. Next, any second choice votes for the top two from voters whose first choice was eliminated are added to the original count. The slate with the most votes wins, or if they are tied, the one with the lowest ID (the one submitted first) wins.

Some new `Gatekeeper` functions of interest include:
- actions:
  - `revealBallot()` / `revealManyBallots()`
  - `countVotes()`
  - `countRunoffVotes()`
- queries:
  - `contestStatus()`
  - `getWinningSlate()`
  - `hasPermission()`

Some new `Slate` functions of interest:
- actions:
  - `markAccepted()`
  - `markRejected()`
- queries:
  - `getRequests()`

### CODE REVIEW

File: governance-contracts/contracts/Gatekeeper.sol [hunk 1]
```diff
@@ -12,7 +12,26 @@ contract Gatekeeper {
     event PermissionRequested(uint requestID, bytes metadataHash);
     event SlateCreated(uint slateID, address indexed recommender, bytes metadataHash);
     event VotingTokensDeposited(address indexed voter, uint numTokens);
-    event BallotCommitted(uint ballotID, address indexed voter, uint numTokens, bytes32 commitHash);
+    event BallotCommitted(uint indexed ballotID, address indexed voter, uint numTokens, bytes32 commitHash);
```
Comment (akuanti) [inline governance-contracts/contracts/Gatekeeper.sol]: @kangarang let me know if there's anything you want from these events that isn't here, or if you think this is overkill.

Comment (kangarang) [reply governance-contracts/contracts/Gatekeeper.sol]: this looks fine for now. when it comes time to improving the api (maybe with an event listener), and/or before the contract audits, let's revisit the event emission args

File: governance-contracts/contracts/Gatekeeper.sol [hunk 2]
```diff
@@ -242,6 +318,366 @@ contract Gatekeeper {
...
+
+        // Update tally for each contest
+        for (uint i = 0; i < votes; i++) {
+            uint category = categories[i];
+
+            // get the contest for the current category
+            Contest storage contest = ballot.contests[category];
+
+            // Increment totals for first and second choice slates
+            uint firstChoice = firstChoices[i];
```
Comment (akuanti) [inline governance-contracts/contracts/Gatekeeper.sol]: Do we need to enforce that `firstChoice != secondChoice`?

File: governance-contracts/test/utils.js [hunk 3]
```diff
@@ -112,6 +115,181 @@ function zeroHash() {
...
 }
 
+
+/**
+ * generateCommitHash
+ * keccak256(category + firstChoice + secondChoice ... + salt)
+ * @param {*} votes { category: { firstChoice, secondChoice }}
+ * @param {ethUtils.BN} salt Random 256-bit number
+ */
+function generateCommitHash(votes, salt) {
```
Comment (kangarang) [inline governance-contracts/test/utils.js]: any particular reason we are not using the one in `panvala-utils`?

File: governance-contracts/contracts/Gatekeeper.sol [hunk 4]
```diff
@@ -111,44 +152,65 @@ contract Gatekeeper {
...
     // SLATE GOVERNANCE
     /**
     * @dev Create a new slate with the associated requestIds and metadata hash.
     * @param batchNumber The batch to submit for
-    * @param category The category to submit the slate for
+    * @param categoryID The category to submit the slate for
     * @param requestIDs A list of request IDs to include in the slate
     * @param metadataHash A reference to metadata about the slate
     */
     function recommendSlate(
```
Comment (kangarang) [inline governance-contracts/contracts/Gatekeeper.sol]: more explicit: `recommendSlateByCategory`

Comment (akuanti) [reply governance-contracts/contracts/Gatekeeper.sol]: There isn't any other way to recommend a slate though. You always need to specify the category.

File: governance-contracts/test/utils.js (hunk 3 above)
Comment (akuanti) [reply governance-contracts/test/utils.js]: Not really. I can change that.

File: governance-contracts/contracts/Gatekeeper.sol (hunk 4 above)
Comment (kangarang) [reply governance-contracts/contracts/Gatekeeper.sol]: good point

File: governance-contracts/contracts/Gatekeeper.sol [hunk 5]
```diff
@@ -242,6 +318,366 @@ contract Gatekeeper {
...
+        uint winnerPercentage = winnerVotes.mul(100).div(total);
+        if (winnerPercentage > 50) {
+            updatedContest.winner = winner;
+            acceptWinningSlate(winner);
+            rejectLosingSlates(ballotID, categoryID);
+            updatedContest.status = ContestStatus.VoteFinalized;
+            emit ConfidenceVoteFinalized(ballotID, categoryID, winner);
+        } else {
+            rejectEliminatedSlates(ballotID, categoryID);
+            updatedContest.status = ContestStatus.RunoffPending;
```
Comment (kangarang) [inline governance-contracts/contracts/Gatekeeper.sol]: emit a notification event such as `RunoffVotePending` or `ConfidenceVoteFailed`

File: governance-contracts/test/gatekeeper.js [hunk 6]
```diff
@@ -439,4 +523,1331 @@ contract('Gatekeeper', (accounts) => {
...
+      // basic confidence vote
+      // slate 0 should win
+
+      // Commit for voters
+      const aliceReveal = await voteSingle(gatekeeper, alice, GRANT, 0, 1, '1000', '1234');
+      const bobReveal = await voteSingle(gatekeeper, bob, GRANT, 0, 1, '1000', '5678');
+      const carolReveal = await voteSingle(gatekeeper, carol, GRANT, 1, 0, '1000', '9012');
+
+      // Reveal all votes
+      await reveal(gatekeeper, aliceReveal);
```
Comment (kangarang) [inline governance-contracts/test/gatekeeper.js]: where is this function defined?

Comment (kangarang) [reply governance-contracts/test/gatekeeper.js]: nvm, found it

**Comment by kangarang** [comment]
this LGTM. excellent work @akuanti!

reference items:
- https://github.com/ConsenSys/panvala/pull/36#discussion_r275013506
- https://github.com/ConsenSys/panvala/pull/36#discussion_r275044467

**Pull request merged**